test/gtests/src/test_Serialisation.cpp

```diff
@@ -58,21 +63,93 @@ TEST_F(Test_DerefInputFnParams, BasicTest) {
 class Test_MatMulInt8Params : public ::testing::Test {};
 
 TEST_F(Test_MatMulInt8Params, BasicTest) {
-  test_serialisation<MatMulInt8::Params>();
+  int8_t d[sizeof(MatMulInt8::Params)];
+
+  for (int t = 0; t < test_count; ++t) {
+    for (int idx = 0; idx < sizeof(d); ++idx) d[idx] = rng.rand<int8_t>();
+
+    MatMulInt8::Params *p = (MatMulInt8::Params *)d;
+
+    p->bytes_per_kernel_channel = (int)rng.rand<uint8_t>();
+    p->output_slice_channel_count = (int)rng.rand<uint8_t>();
+
+    int weight_count = MatMulInt8::get_weights_bytes(
+        p->bytes_per_kernel_channel, p->output_slice_channel_count);
+
+    std::vector<int8_t> weights(weight_count);
+
+    for (int idx = 0; idx < weight_count; ++idx)
+      weights[idx] = rng.rand<int8_t>();
+
+    p->weights = (int8_t *)weights.data();
+
+    std::string s = p->template serialise<MatMulInt8::Params>();
+
+    int allocation_byte_count = p->get_allocation_byte_count();
+    char *allocated_memory = (char *)std::malloc(allocation_byte_count);
+
+    MatMulInt8::Params *q = MatMulInt8::Params::deserialise<MatMulInt8::Params>(
+        allocated_memory, s.c_str());
+
+    EXPECT_NE(p, q);
+    EXPECT_EQ(p->output_slice_channel_count, q->output_slice_channel_count);
+    EXPECT_EQ(p->bytes_per_kernel_channel, q->bytes_per_kernel_channel);
+
+    for (size_t i = 0; i < weight_count; i++) {
+      EXPECT_EQ(p->weights[i], q->weights[i]);
+    }
+    free(allocated_memory);
+  }
 }
 
 class Test_MatMulDirectFnParams : public ::testing::Test {};
 
 TEST_F(Test_MatMulDirectFnParams, BasicTest) {
-  test_serialisation<MatMulDirectFn::Params>();
+  int8_t d[sizeof(MatMulDirectFn::Params)];
+
+  for (int t = 0; t < test_count; ++t) {
+    int weight_bytes = t;
+
+    for (int idx = 0; idx < sizeof(d); ++idx) d[idx] = rng.rand<int8_t>();
+
+    std::vector<int8_t> weights(weight_bytes);
+
+    for (int idx = 0; idx < weight_bytes; ++idx)
+      weights[idx] = rng.rand<int8_t>();
+
+    MatMulDirectFn::Params *p = (MatMulDirectFn::Params *)d;
+    p->weights = (int8_t *)weights.data();
+    p->weights_bytes = weight_bytes;
+
+    std::string s = p->template serialise<MatMulDirectFn::Params>();
+
+    int allocation_byte_count = p->get_allocation_byte_count();
```

Comment:
How would this work on the deserialisation side since we wouldn't have access to this function? Can it be made part of the serialized string such as the first four bytes, so we can directly access it?